File: typescript/src/code/code.ts
import {
  CodeExecutionResult,
  CodeExecutionResultItem,
  CodeExecutionLogs,
  CodeExecutionError,
} from "../types/api-response";

/**
 * Handles code execution operations in the AgentBay cloud environment.
 */
export class Code {
  private session: {
    getAPIKey(): string;
    getSessionId(): string;
    getWsClient?: () => Promise<any>;
    mcpTools?: Array<{ name?: string; server?: string }>;
    callMcpTool(
      toolName: string,
      args: any,
      autoGenSession?: boolean
    ): Promise<{
      success: boolean;
      data: string;
      errorMessage: string;
      requestId: string;
    }>;
  };

  /**
   * Initialize a Code object.
   *
   * @param session - The Session instance that this Code belongs to.
   */
  constructor(session: {
    getAPIKey(): string;
    getSessionId(): string;
    getWsClient?: () => Promise<any>;
    mcpTools?: Array<{ name?: string; server?: string }>;
    callMcpTool(
      toolName: string,
      args: any,
      autoGenSession?: boolean
    ): Promise<{
      success: boolean;
      data: string;
      errorMessage: string;
      requestId: string;
    }>;
  }) {
    this.session = session;
  }

  /**
   * Parses the backend JSON response into a structured CodeExecutionResult
   * @param data Raw JSON string from backend
   */
  private parseBackendResponse(data: string): CodeExecutionResult {
    let raw: any;
    try {
      raw = JSON.parse(data);
      // Handle double-encoded JSON if necessary
      if (typeof raw === 'string') {
        try {
          raw = JSON.parse(raw);
        } catch (e) {
          // If second parse fails, keep as string (legacy output)
        }
      }
    } catch (e) {
      // Not JSON, assume legacy plain string result or simple error message
      return {
        requestId: "", // Will be filled later
        success: true,
        result: data,
        logs: { stdout: [data], stderr: [] },
        results: [{ text: data, isMainResult: true }],
      };
    }

    // Check if it's the expected structure (has result array or logs)
    // If parsed but not the structure we want, treat as plain object result?
    // But usually if it parses as object, it's likely our structure or empty object.
    
    const logs: CodeExecutionLogs = {
      stdout: raw.stdout || [],
      stderr: raw.stderr || [],
    };

    const results: CodeExecutionResultItem[] = [];
    if (Array.isArray(raw.result)) {
      for (const itemStr of raw.result) {
        try {
          // Items in 'result' are JSON strings themselves
          const itemMap = typeof itemStr === 'string' ? JSON.parse(itemStr) : itemStr;
          const item: CodeExecutionResultItem = {};

          if (itemMap.isMainResult) item.isMainResult = true;
          if (itemMap.is_main_result) item.isMainResult = true;

          // Map MIME types to fields
          if (itemMap["text/plain"]) item.text = itemMap["text/plain"];
          if (itemMap["text/html"]) item.html = itemMap["text/html"];
          if (itemMap["text/markdown"]) item.markdown = itemMap["text/markdown"];
          if (itemMap["image/png"]) item.png = itemMap["image/png"];
          if (itemMap["image/jpeg"]) item.jpeg = itemMap["image/jpeg"];
          if (itemMap["image/svg+xml"]) item.svg = itemMap["image/svg+xml"];
          if (itemMap["text/latex"]) item.latex = itemMap["text/latex"];
          if (itemMap["application/json"]) item.json = itemMap["application/json"];

          // Chart support (Vega/Vega-Lite)
          if (itemMap["application/vnd.vegalite.v4+json"]) {
            item.chart = itemMap["application/vnd.vegalite.v4+json"];
          } else if (itemMap["application/vnd.vegalite.v5+json"]) {
            item.chart = itemMap["application/vnd.vegalite.v5+json"];
          } else if (itemMap["application/vnd.vega.v5+json"]) {
            item.chart = itemMap["application/vnd.vega.v5+json"];
          }

          results.push(item);
        } catch (e) {
          // Ignore parse errors for individual items
        }
      }
    }

    let error: CodeExecutionError | undefined;
    if (raw.executionError) {
      error = {
        name: "ExecutionError",
        value: raw.executionError,
        traceback: "",
      };
    }

    // Determine legacy result string for backward compatibility
    let resultText = "";
    const mainRes = results.find((r) => r.isMainResult && r.text);
    if (mainRes && mainRes.text) {
      resultText = mainRes.text;
    } else if (results.length > 0 && results[0].text) {
      resultText = results[0].text;
    } else if (logs.stdout.length > 0) {
      resultText = logs.stdout.join("");
    }

    return {
      requestId: "", // Will be filled by caller
      success: !raw.executionError,
      result: resultText,
      errorMessage: raw.executionError,
      logs,
      results,
      error,
      executionTime: raw.executionTime || raw.execution_time,
      executionCount: raw.executionCount || raw.execution_count,
    };
  }

  private safeToString(v: any): string {
    try {
      if (typeof v === "string") return v;
      return JSON.stringify(v);
    } catch (_e) {
      return String(v);
    }
  }

  private toResultItem(payload: any): CodeExecutionResultItem {
    const item: CodeExecutionResultItem = {};
    if (!payload || typeof payload !== "object") {
      item.text = String(payload);
      return item;
    }

    if (payload.isMainResult) item.isMainResult = true;
    if (payload.is_main_result) item.isMainResult = true;

    if (payload["text/plain"]) item.text = payload["text/plain"];
    if (payload["text/html"]) item.html = payload["text/html"];
    if (payload["text/markdown"]) item.markdown = payload["text/markdown"];
    if (payload["image/png"]) item.png = payload["image/png"];
    if (payload["image/jpeg"]) item.jpeg = payload["image/jpeg"];
    if (payload["image/svg+xml"]) item.svg = payload["image/svg+xml"];
    if (payload["text/latex"]) item.latex = payload["text/latex"];
    if (payload["application/json"]) item.json = payload["application/json"];

    if (payload["application/vnd.vegalite.v4+json"]) {
      item.chart = payload["application/vnd.vegalite.v4+json"];
    } else if (payload["application/vnd.vegalite.v5+json"]) {
      item.chart = payload["application/vnd.vegalite.v5+json"];
    } else if (payload["application/vnd.vega.v5+json"]) {
      item.chart = payload["application/vnd.vega.v5+json"];
    }

    if (payload.text && !item.text) item.text = payload.text;
    return item;
  }

  private async runCodeStreamWs(params: {
    code: string;
    language: string;
    timeoutS: number;
    onStdout?: (chunk: string) => void;
    onStderr?: (chunk: string) => void;
    onError?: (err: any) => void;
  }): Promise<CodeExecutionResult> {
    const stdoutChunks: string[] = [];
    const stderrChunks: string[] = [];
    const results: CodeExecutionResultItem[] = [];
    let errorObj: CodeExecutionError | undefined;
    let errorMessage = "";

    // Determine target from MCP tool list if available.
    let target = "wuying_codespace";
    for (const tool of this.session.mcpTools || []) {
      if (tool && tool.name === "run_code" && tool.server) {
        target = tool.server;
        break;
      }
    }

    if (!this.session.getWsClient) {
      return {
        requestId: "",
        success: false,
        result: "",
        errorMessage: "WS streaming is not available in this session",
      };
    }

    const wsClient = await this.session.getWsClient();

    const handle = await wsClient.callStream({
      target,
      data: {
        method: "run_code",
        mode: "stream",
        params: {
          language: params.language,
          timeoutS: params.timeoutS,
          code: params.code,
        },
      },
      onEvent: (_invocationId: string, data: any) => {
        const eventType = data?.eventType;
        if (eventType === "stdout") {
          const chunk = String(data?.chunk ?? "");
          stdoutChunks.push(chunk);
          if (params.onStdout) params.onStdout(chunk);
          return;
        }
        if (eventType === "stderr") {
          const chunk = String(data?.chunk ?? "");
          stderrChunks.push(chunk);
          if (params.onStderr) params.onStderr(chunk);
          return;
        }
        if (eventType === "result") {
          const r = this.toResultItem(data?.result);
          results.push(r);
          return;
        }
        if (eventType === "error") {
          const msg =
            typeof data?.error === "string" ? data.error : this.safeToString(data);
          errorMessage = msg;
          errorObj = {
            name: "ExecutionError",
            value: msg,
            traceback: "",
          };
          if (params.onError) params.onError(data);
          return;
        }
      },
      onEnd: (_invocationId: string, data: any) => {
        const executionError = data?.executionError;
        if (executionError && !errorObj) {
          const msg = String(executionError);
          errorMessage = msg;
          errorObj = { name: "ExecutionError", value: msg, traceback: "" };
        }
      },
      onError: (_invocationId: string, err: any) => {
        const msg = err instanceof Error ? err.message : String(err);
        errorMessage = msg;
        if (!errorObj) {
          errorObj = { name: "ExecutionError", value: msg, traceback: "" };
        }
        if (params.onError) params.onError(err);
      },
    });

    let endData: any = {};
    try {
      endData = await handle.waitEnd();
    } catch (e: any) {
      const msg = e instanceof Error ? e.message : String(e);
      return {
        requestId: handle.invocationId,
        success: false,
        result: "",
        errorMessage: msg,
        logs: { stdout: stdoutChunks, stderr: stderrChunks },
        results,
        error: errorObj,
      };
    }

    const ok = !errorObj && !endData?.executionError && endData?.status !== "failed";

    const logs: CodeExecutionLogs = {
      stdout: stdoutChunks,
      stderr: stderrChunks,
    };

    // Determine legacy result string for backward compatibility
    let resultText = "";
    const mainRes = results.find((r) => r.isMainResult && r.text);
    if (mainRes && mainRes.text) {
      resultText = mainRes.text;
    } else if (results.length > 0 && results[0].text) {
      resultText = results[0].text;
    } else if (logs.stdout.length > 0) {
      resultText = logs.stdout.join("");
    }

    return {
      requestId: handle.invocationId,
      success: ok,
      result: resultText,
      errorMessage: ok ? "" : errorMessage || String(endData?.executionError || ""),
      logs,
      results,
      error: errorObj,
      executionTime: endData?.executionTime,
      executionCount: endData?.executionCount,
    };
  }

  /**
   * Execute code in the specified language with a timeout.
   * Corresponds to Python's run_code() method
   *
   * @param code - The code to execute.
   * @param language - The programming language of the code. Case-insensitive.
   *                  Supported values: 'python', 'javascript', 'r', 'java'.
   * @param timeoutS - The timeout for the code execution in seconds. Default is 60s.
   *                   Note: Due to gateway limitations, each request cannot exceed 60 seconds.
   * @returns CodeExecutionResult with code execution output and requestId
   * @throws Error if an unsupported language is specified.
   *
   * @example
   * ```typescript
   * const agentBay = new AgentBay({ apiKey: 'your_api_key' });
   * const result = await agentBay.create({ imageId: "code_latest" });
   * if (result.success) {
   *   const codeResult = await result.session.code.runCode('print("Hello")', "python");
   *   console.log(codeResult.result);
   *   if (codeResult.results) {
   *      // Access rich output like images or charts
   *   }
   *   await result.session.delete();
   * }
   * ```
   */
  async runCode(
    code: string,
    language: string,
    timeoutS = 60
  ): Promise<CodeExecutionResult> {
    try {
      // Normalize and validate language (case-insensitive)
      const rawLanguage = language ?? "";
      const normalizedLanguage = String(rawLanguage).trim().toLowerCase();
      const aliases: Record<string, string> = {
        py: "python",
        python3: "python",
        js: "javascript",
        node: "javascript",
        nodejs: "javascript",
      };
      const canonicalLanguage = aliases[normalizedLanguage] || normalizedLanguage;
      const supported = new Set(["python", "javascript", "r", "java"]);

      if (!supported.has(canonicalLanguage)) {
        return {
          requestId: "",
          success: false,
          result: "",
          errorMessage: `Unsupported language: ${rawLanguage}. Supported languages are 'python', 'javascript', 'r', and 'java'`,
        };
      }

      // Streaming is temporarily disabled in this version.
      // The streaming implementation is preserved in runCodeStreamWs()
      // and will be re-enabled in a future release.

      const args = {
        code,
        language: canonicalLanguage,
        timeout_s: timeoutS,
      };

      const response = await this.session.callMcpTool("run_code", args, false);

      let codeResult: CodeExecutionResult;
      
      // Even if response.success is false, data might contain structured error info
      if (response.success) {
        codeResult = this.parseBackendResponse(response.data);
      } else {
        // Check if errorMessage is JSON
        try {
            if (response.errorMessage && response.errorMessage.trim().startsWith('{')) {
                codeResult = this.parseBackendResponse(response.errorMessage);
                codeResult.success = false;
            } else {
                 // Real failure not from code execution but tool failure
                 return {
                    requestId: response.requestId,
                    success: false,
                    result: "",
                    errorMessage: response.errorMessage,
                  };
            }
        } catch {
             return {
                requestId: response.requestId,
                success: false,
                result: "",
                errorMessage: response.errorMessage,
              };
        }
      }

      codeResult.requestId = response.requestId;
      return codeResult;

    } catch (error) {
      return {
        requestId: "",
        success: false,
        result: "",
        errorMessage: `Failed to run code: ${error}`,
      };
    }
  }

  /**
   * Alias of runCode().
   */
  async run(
    code: string,
    language: string,
    timeoutS = 60
  ): Promise<CodeExecutionResult> {
    return await this.runCode(code, language, timeoutS);
  }

  /**
   * Alias of runCode().
   */
  async execute(
    code: string,
    language: string,
    timeoutS = 60
  ): Promise<CodeExecutionResult> {
    return await this.runCode(code, language, timeoutS);
  }
}

File: golang/pkg/agentbay/code/code.go
package code

import (
	"encoding/json"
	"fmt"
	"strings"

	mcp "github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/internal"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
)

// WsStreamClient is a public interface for WebSocket streaming clients.
// It abstracts the internal WsClient to allow external packages (e.g., tests)
// to implement GetWsClient without importing the internal package.
type WsStreamClient interface {
	CallStream(target string, data map[string]interface{},
		onEvent func(string, map[string]interface{}),
		onEnd func(string, map[string]interface{}),
		onError func(string, error),
	) (*internal.WsStreamHandle, error)
}

// CodeExecutionError represents an error during code execution
type CodeExecutionError struct {
	Name      string `json:"name"`
	Value     string `json:"value"`
	Traceback string `json:"traceback"`
}

// CodeExecutionLogs represents stdout and stderr logs
type CodeExecutionLogs struct {
	Stdout []string `json:"stdout"`
	Stderr []string `json:"stderr"`
}

// CodeExecutionResultItem represents a single result item (text, image, etc.)
type CodeExecutionResultItem struct {
	Text         string      `json:"text,omitempty"`
	HTML         string      `json:"html,omitempty"`
	Markdown     string      `json:"markdown,omitempty"`
	PNG          string      `json:"png,omitempty"`
	JPEG         string      `json:"jpeg,omitempty"`
	SVG          string      `json:"svg,omitempty"`
	Latex        string      `json:"latex,omitempty"`
	JSON         interface{} `json:"json,omitempty"`
	Chart        interface{} `json:"chart,omitempty"`
	IsMainResult bool        `json:"is_main_result,omitempty"`
}

// CodeResult represents the result of a code execution
type CodeResult struct {
	models.ApiResponse // Embedded ApiResponse

	// Legacy/Simple output
	Output string

	// Enhanced fields
	Success        bool                      `json:"success"`
	Result         string                    `json:"result"` // Legacy compatible result text
	ErrorMessage   string                    `json:"error_message,omitempty"`
	Logs           *CodeExecutionLogs        `json:"logs,omitempty"`
	Results        []CodeExecutionResultItem `json:"results,omitempty"`
	Error          *CodeExecutionError       `json:"error,omitempty"`
	ExecutionTime  float64                   `json:"execution_time,omitempty"`
	ExecutionCount *int                      `json:"execution_count,omitempty"`
}

// Code handles code execution operations in the AgentBay cloud environment.
type Code struct {
	Session interface {
		GetAPIKey() string
		GetClient() *mcp.Client
		GetSessionId() string
		CallMcpTool(toolName string, args interface{}) (*models.McpToolResult, error)
		GetWsClient() (interface{}, error)
	}
}

// NewCode creates a new Code instance
func NewCode(session interface {
	GetAPIKey() string
	GetClient() *mcp.Client
	GetSessionId() string
	CallMcpTool(toolName string, args interface{}) (*models.McpToolResult, error)
	GetWsClient() (interface{}, error)
}) *Code {
	return &Code{
		Session: session,
	}
}

type RunCodeStreamBetaOptions struct {
	TimeoutS   int
	StreamBeta bool
	OnStdout   func(chunk string)
	OnStderr   func(chunk string)
	OnError    func(err error)
}

func (c *Code) runCodeStreamWs(code string, language string, timeoutS int, opts *RunCodeStreamBetaOptions) (*CodeResult, error) {
	if timeoutS <= 0 {
		timeoutS = 60
	}

	rawLanguage := language
	normalizedLanguage := strings.ToLower(strings.TrimSpace(language))
	aliases := map[string]string{
		"py":      "python",
		"python3": "python",
		"js":      "javascript",
		"node":    "javascript",
		"nodejs":  "javascript",
	}
	if canonical, ok := aliases[normalizedLanguage]; ok {
		normalizedLanguage = canonical
	}

	switch normalizedLanguage {
	case "python", "javascript", "r", "java":
	default:
		return &CodeResult{
			Success:      false,
			ErrorMessage: fmt.Sprintf("Unsupported language: %s. Supported languages are 'python', 'javascript', 'r', and 'java'", rawLanguage),
		}, nil
	}

	wsClientRaw, err := c.Session.GetWsClient()
	if err != nil {
		return &CodeResult{
			Success:      false,
			ErrorMessage: err.Error(),
		}, nil
	}
	wsClient, wsOk := wsClientRaw.(*internal.WsClient)
	if !wsOk || wsClient == nil {
		return &CodeResult{
			Success:      false,
			ErrorMessage: "invalid or nil WsClient returned from session",
		}, nil
	}

	stdoutChunks := []string{}
	stderrChunks := []string{}
	results := []CodeExecutionResultItem{}
	var executionCount *int
	executionTime := float64(0)
	var execErrMsg string

	target := "wuying_codespace"
	handle, err := wsClient.CallStream(
		target,
		map[string]interface{}{
			"method": "run_code",
			"mode":   "stream",
			"params": map[string]interface{}{
				"language": normalizedLanguage,
				"timeoutS": timeoutS,
				"code":     code,
			},
		},
		func(_ string, data map[string]interface{}) {
			eventType, _ := data["eventType"].(string)
			switch eventType {
			case "stdout":
				chunk := fmt.Sprintf("%v", data["chunk"])
				stdoutChunks = append(stdoutChunks, chunk)
				if opts != nil && opts.OnStdout != nil {
					opts.OnStdout(chunk)
				}
			case "stderr":
				chunk := fmt.Sprintf("%v", data["chunk"])
				stderrChunks = append(stderrChunks, chunk)
				if opts != nil && opts.OnStderr != nil {
					opts.OnStderr(chunk)
				}
			case "result":
				if m, ok := data["result"].(map[string]interface{}); ok {
					results = append(results, parseResultItemFromMimeMap(m))
				} else {
					results = append(results, CodeExecutionResultItem{Text: fmt.Sprintf("%v", data["result"])})
				}
			case "error":
				execErrMsg = fmt.Sprintf("%v", data["error"])
				if opts != nil && opts.OnError != nil {
					opts.OnError(fmt.Errorf("%s", execErrMsg))
				}
			}
		},
		func(_ string, data map[string]interface{}) {
			if v, ok := data["executionCount"]; ok {
				if i, ok := toInt(v); ok {
					executionCount = &i
				}
			}
			if v, ok := data["executionTime"]; ok {
				if f, ok := toFloat(v); ok {
					executionTime = f
				}
			}
			if v, ok := data["executionError"]; ok && v != nil && fmt.Sprintf("%v", v) != "" {
				execErrMsg = fmt.Sprintf("%v", v)
			}
		},
		func(_ string, e error) {
			if opts != nil && opts.OnError != nil {
				opts.OnError(e)
			}
			if execErrMsg == "" {
				execErrMsg = e.Error()
			}
		},
	)
	if err != nil {
		return &CodeResult{
			ApiResponse:  models.ApiResponse{RequestID: ""},
			Success:      false,
			ErrorMessage: err.Error(),
		}, nil
	}

	endData, err := handle.WaitEnd()
	if err != nil {
		return &CodeResult{
			ApiResponse:  models.ApiResponse{RequestID: handle.InvocationID},
			Success:      false,
			ErrorMessage: err.Error(),
			Logs:         &CodeExecutionLogs{Stdout: stdoutChunks, Stderr: stderrChunks},
			Results:      results,
		}, nil
	}
	_ = endData

	ok := execErrMsg == ""
	if st, okSt := endData["status"].(string); okSt && st == "failed" {
		ok = false
	}

	res := &CodeResult{
		ApiResponse:    models.ApiResponse{RequestID: handle.InvocationID},
		Success:        ok,
		ErrorMessage:   execErrMsg,
		Logs:           &CodeExecutionLogs{Stdout: stdoutChunks, Stderr: stderrChunks},
		Results:        results,
		ExecutionTime:  executionTime,
		ExecutionCount: executionCount,
	}

	foundText := false
	for _, r := range res.Results {
		if r.IsMainResult && r.Text != "" {
			res.Output = r.Text
			res.Result = r.Text
			foundText = true
			break
		}
	}
	if !foundText && len(res.Results) > 0 && res.Results[0].Text != "" {
		res.Output = res.Results[0].Text
		res.Result = res.Results[0].Text
		foundText = true
	}
	if !foundText && len(stdoutChunks) > 0 {
		res.Output = strings.Join(stdoutChunks, "")
		res.Result = res.Output
	}

	if !res.Success && res.ErrorMessage == "" {
		res.ErrorMessage = "Failed to run code"
	}
	return res, nil
}

func parseResultItemFromMimeMap(itemMap map[string]interface{}) CodeExecutionResultItem {
	item := CodeExecutionResultItem{
		JSON:  itemMap["application/json"],
		Chart: itemMap["chart"],
	}

	if val, ok := itemMap["isMainResult"].(bool); ok {
		item.IsMainResult = val
	}
	if val, ok := itemMap["is_main_result"].(bool); ok {
		item.IsMainResult = val
	}
	if val, ok := itemMap["text/plain"].(string); ok {
		item.Text = val
	}
	if val, ok := itemMap["text/html"].(string); ok {
		item.HTML = val
	}
	if val, ok := itemMap["text/markdown"].(string); ok {
		item.Markdown = val
	}
	if val, ok := itemMap["image/png"].(string); ok {
		item.PNG = val
	}
	if val, ok := itemMap["image/jpeg"].(string); ok {
		item.JPEG = val
	}
	if val, ok := itemMap["image/svg+xml"].(string); ok {
		item.SVG = val
	}
	if val, ok := itemMap["text/latex"].(string); ok {
		item.Latex = val
	}
	if val, ok := itemMap["application/vnd.vegalite.v4+json"]; ok {
		item.Chart = val
	}
	if val, ok := itemMap["application/vnd.vegalite.v5+json"]; ok {
		item.Chart = val
	}
	if val, ok := itemMap["application/vnd.vega.v5+json"]; ok {
		item.Chart = val
	}
	return item
}

func toInt(v interface{}) (int, bool) {
	switch t := v.(type) {
	case float64:
		return int(t), true
	case int:
		return t, true
	case int64:
		return int(t), true
	case json.Number:
		i, err := t.Int64()
		if err != nil {
			return 0, false
		}
		return int(i), true
	default:
		return 0, false
	}
}

func toFloat(v interface{}) (float64, bool) {
	switch t := v.(type) {
	case float64:
		return t, true
	case int:
		return float64(t), true
	case int64:
		return float64(t), true
	case json.Number:
		f, err := t.Float64()
		if err != nil {
			return 0, false
		}
		return f, true
	default:
		return 0, false
	}
}

// backendResponse represents the raw JSON structure returned by the backend tool
type backendResponse struct {
	ExecutionError string   `json:"executionError"`
	Result         []string `json:"result"` // List of JSON strings
	Stdout         []string `json:"stdout"`
	Stderr         []string `json:"stderr"`
	TraceID        string   `json:"traceId"`
	// Try both cases just in case
	ExecutionTime   float64 `json:"executionTime"`
	ExecutionTimeSn float64 `json:"execution_time"`
	ExecutionCount  *int    `json:"executionCount"`
	ExecutionCntSn  *int    `json:"execution_count"`
}

func parseBackendResponse(data string) (*CodeResult, error) {
	var raw backendResponse
	if err := json.Unmarshal([]byte(data), &raw); err != nil {
		return nil, err
	}

	// Check if it's actually the expected structure (e.g. has result or stdout/stderr keys)
	// If unmarshal into empty struct, it might succeed but be empty.
	// But since Result/Stdout are slices, they will be nil or empty.
	// We can check if data looks like JSON object.
	// But let's assume if Unmarshal works and we find keys, it's good.

	logs := &CodeExecutionLogs{
		Stdout: raw.Stdout,
		Stderr: raw.Stderr,
	}
	if logs.Stdout == nil {
		logs.Stdout = []string{}
	}
	if logs.Stderr == nil {
		logs.Stderr = []string{}
	}

	execTime := raw.ExecutionTime
	if execTime == 0 {
		execTime = raw.ExecutionTimeSn
	}

	execCount := raw.ExecutionCount
	if execCount == nil {
		execCount = raw.ExecutionCntSn
	}

	res := &CodeResult{
		Success:        true,
		Logs:           logs,
		ExecutionTime:  execTime,
		ExecutionCount: execCount,
	}

	// Parse Results
	for _, resultStr := range raw.Result {
		var itemMap map[string]interface{}
		if err := json.Unmarshal([]byte(resultStr), &itemMap); err != nil {
			continue
		}

		item := CodeExecutionResultItem{
			JSON:  itemMap["json"],
			Chart: itemMap["chart"],
		}

		if val, ok := itemMap["isMainResult"].(bool); ok {
			item.IsMainResult = val
		}

		// Map MIME types to fields
		if val, ok := itemMap["text/plain"].(string); ok {
			item.Text = val
		}
		if val, ok := itemMap["text/html"].(string); ok {
			item.HTML = val
		}
		if val, ok := itemMap["text/markdown"].(string); ok {
			item.Markdown = val
		}
		if val, ok := itemMap["image/png"].(string); ok {
			item.PNG = val
		}
		if val, ok := itemMap["image/jpeg"].(string); ok {
			item.JPEG = val
		}
		if val, ok := itemMap["image/svg+xml"].(string); ok {
			item.SVG = val
		}
		if val, ok := itemMap["text/latex"].(string); ok {
			item.Latex = val
		}
		// Handle JSON and Chart mime types
		if val, ok := itemMap["application/json"]; ok {
			item.JSON = val
		}
		if val, ok := itemMap["application/vnd.vegalite.v4+json"]; ok {
			item.Chart = val
		}
		if val, ok := itemMap["application/vnd.vegalite.v5+json"]; ok {
			item.Chart = val
		}
		if val, ok := itemMap["application/vnd.vega.v5+json"]; ok {
			item.Chart = val
		}

		res.Results = append(res.Results, item)
	}
	if res.Results == nil {
		res.Results = []CodeExecutionResultItem{}
	}

	// Handle Error
	if raw.ExecutionError != "" {
		res.Success = false
		res.Error = &CodeExecutionError{
			Value: raw.ExecutionError,
			// Name and Traceback might not be parsed from simple string
		}
		res.ErrorMessage = raw.ExecutionError
	}

	// Construct Result/Output string for backward compatibility
	foundText := false
	for _, r := range res.Results {
		if r.IsMainResult && r.Text != "" {
			res.Output = r.Text
			res.Result = r.Text
			foundText = true
			break
		}
	}
	if !foundText && len(res.Results) > 0 && res.Results[0].Text != "" {
		res.Output = res.Results[0].Text
		res.Result = res.Results[0].Text
		foundText = true
	}
	if !foundText && len(res.Logs.Stdout) > 0 {
		res.Output = strings.Join(res.Logs.Stdout, "")
		res.Result = res.Output
	}

	return res, nil
}

// RunCode executes code in the session environment.
// timeoutS: The timeout for the code execution in seconds. Default is 60s.
// Note: Due to gateway limitations, each request cannot exceed 60 seconds.
//
// language is case-insensitive. Supported values: "python", "javascript", "r", "java".
//
// Example:
//
//	client, _ := agentbay.NewAgentBay(os.Getenv("AGENTBAY_API_KEY"), nil)
//	sessionResult, _ := client.Create(agentbay.NewCreateSessionParams().WithImageId("code_latest"))
//	defer sessionResult.Session.Delete()
//	codeResult, _ := sessionResult.Session.Code.RunCode("print('Hello')", "python")
func (c *Code) RunCode(code string, language string, args ...interface{}) (*CodeResult, error) {
	timeout := 60
	var opts *RunCodeStreamBetaOptions
	for _, a := range args {
		if a == nil {
			continue
		}
		switch v := a.(type) {
		case int:
			if v > 0 {
				timeout = v
			}
		case int32:
			if v > 0 {
				timeout = int(v)
			}
		case int64:
			if v > 0 {
				timeout = int(v)
			}
		case RunCodeStreamBetaOptions:
			tmp := v
			opts = &tmp
		case *RunCodeStreamBetaOptions:
			opts = v
		default:
			return nil, fmt.Errorf("unsupported RunCode argument type: %T", a)
		}
	}

	if opts != nil && opts.TimeoutS > 0 {
		timeout = opts.TimeoutS
	}

	useStream := false
	if opts != nil && (opts.StreamBeta || opts.OnStdout != nil || opts.OnStderr != nil || opts.OnError != nil) {
		useStream = true
	}
	if useStream {
		return c.runCodeStreamWs(code, language, timeout, opts)
	}

	// Normalize and validate language (case-insensitive)
	rawLanguage := language
	normalizedLanguage := strings.ToLower(strings.TrimSpace(language))
	aliases := map[string]string{
		"py":      "python",
		"python3": "python",
		"js":      "javascript",
		"node":    "javascript",
		"nodejs":  "javascript",
	}
	if canonical, ok := aliases[normalizedLanguage]; ok {
		normalizedLanguage = canonical
	}

	switch normalizedLanguage {
	case "python", "javascript", "r", "java":
		// supported
	default:
		return nil, fmt.Errorf(
			"unsupported language: %s. Supported languages are 'python', 'javascript', 'r', and 'java'",
			rawLanguage,
		)
	}

	toolArgs := map[string]interface{}{
		"code":      code,
		"language":  normalizedLanguage,
		"timeout_s": timeout,
	}

	// Use Session's CallMcpTool method
	result, err := c.Session.CallMcpTool("run_code", toolArgs)
	if err != nil {
		return nil, fmt.Errorf("failed to execute code: %w", err)
	}

	var codeRes *CodeResult
	var parseErr error

	if result.Success {
		codeRes, parseErr = parseBackendResponse(result.Data)
	} else {
		// Try parsing error message if it looks like JSON
		if strings.HasPrefix(strings.TrimSpace(result.ErrorMessage), "{") {
			codeRes, parseErr = parseBackendResponse(result.ErrorMessage)
			if codeRes != nil {
				codeRes.Success = false
			}
		} else {
			// Not JSON error, real failure
			return nil, fmt.Errorf("code execution failed: %s", result.ErrorMessage)
		}
	}

	// If parsing failed or result was not enhanced
	if parseErr != nil || codeRes == nil {
		if !result.Success {
			return nil, fmt.Errorf("code execution failed: %s", result.ErrorMessage)
		}
		// Legacy fallback
		codeRes = &CodeResult{
			Success: true,
			Output:  result.Data,
			Result:  result.Data,
		}
	}

	codeRes.RequestID = result.RequestID
	codeRes.ApiResponse = models.ApiResponse{RequestID: result.RequestID}

	return codeRes, nil
}

// Run is an alias of RunCode.
func (c *Code) Run(code string, language string, args ...interface{}) (*CodeResult, error) {
	return c.RunCode(code, language, args...)
}

// Execute is an alias of RunCode.
func (c *Code) Execute(code string, language string, args ...interface{}) (*CodeResult, error) {
	return c.RunCode(code, language, args...)
}
